### PR TITLE
New login flow

### PR DESCRIPTION
**Pull request by DoctorMcKay**
TODO:

- Captchas?
- Make sure that incorrect credentials are handled sanely
- QR login eventually?
- Possible to switch to this in a backwards compatible manner?
- Verify how the mobile app works wrt 2FA enable/disable

# Notes

- Differences between JWT refresh and access tokens
    - refresh
        - iss = `"steam"`
        - aud = `["web", "renew", "derive", "mobile"]`
        - Expires ~210 days after issuance
        - per = 1
    - access
        - iss = `"r:<jti of refresh token>"`
        - aud = `["web"]` (for web logins) or `["web", "mobile"]` (for mobile logins)
        - Expires ~24 hours after issuance
        - per = 0
- Mobile app is authorizing with the `access_token` present in the final `PollAuthSessionStatus` response
    - Mobile app auth just sticks the access token in the URL as an `access_token` query parameter, e.g. `/v1?access_token=eyAid...` (even for post requests)
- QR login will not be present in initial release, but should follow soon after
    - I haven't successfully executed a QR login yet, but I believe this to be what's necessary:
        - Auth with an access token with `"mobile"` aud
        - Signature should be hmac-sha256 with shared_secret or identity_secret (suspect shared_secret), signing 14 bytes containing version from QR code (16-bit), client_id from QR code (64-bit), steamid (64-bit)
    - QR URL format is `https://s.team/q/<version>/<client_id>`
    - GetAuthSessionInfo is called before authorizing the session to retrieve location details to present to the user, but this is likely not required
    - PollAuthSessionStatus returns `had_remote_interaction = true` with no other fields filled in after GetAuthSessionInfo is executed

### CODE REVIEW

File: components/login.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+		let startSessionResponse = await this._loginApiReq('POST', 'BeginAuthSessionViaCredentials', {
+			device_friendly_name: SteamCommunity.USER_AGENT,
+			account_name: details.accountName,
+			encrypted_password: hex2b64(key.encrypt(details.password)), // dunno why valve is sending this as a base64 string and not bytes...
+			encryption_timestamp: timestamp,
+			remember_login: true,
+			platform_type: EPlatformType.Win64,
+			persistence: 1,
+			website_id: 'Community'
```
Comment (Revadike) [inline components/login.js:46]: Use enum maybe?

File: components/login.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	let httpReqOptions = {
+		method: httpMethod,
+		uri: `https://api.steampowered.com/IAuthenticationService/${apiMethod}/v1`,
+		headers: API_HEADERS,
+		encoding: null
+	};
+
+	if (inputData) {
+		let proto = Protos[`CAuthentication_${apiMethod}_Request`];
+		let inputBuffer = proto.encode(inputData).finish();
```
Comment (xPaw) [inline components/login.js:191]: You don't have to use `input_protobuf_encoded`, web apis accept plain old form submits just fine, and return json data.

Comment (DoctorMcKay) [reply components/login.js:191]: I'm aware. The intent is to mimic official behavior as closely as possible.

Comment (Revadike) [reply components/login.js:191]: Interesting. I do think it's a good idea to at least write a comment, so people looking at these functions will know about it.

**Comment by DoctorMcKay** [comment]
For anyone watching this PR, https://github.com/DoctorMcKay/node-steam-session is where everything's happening.

**Pull request closed without merging**